File: PythonCode/ChapterQtech/Exercise_6_4_6.py
"""
 This script sets out to fix the duration of the spin-spin interaction 
 between to spin 1/2-particles so that it, in effect, implements the SWAP 
 gate. The parameters are rigged so that the Hamiltonian assumes a 
 particularly simple form. The only inputs are the upper limit for the 
 interaction time T - and dT, the resolution in the vector with 
 time-durations.
 
 It plots the cost function 
 C(T) = 1 - |1/4 * Tr(U_target^\dagger U(T))|^2
 as a function of duration T.
 When C=0, our gate coincides with the target gate, and the fidelity is 100%.
"""

# Libraries
import numpy as np
from matplotlib import pyplot as plt
from scipy import linalg

# Input
Tmax = 3
dT = 0.01

# Vector with durations
Tvector = np.arange(0, Tmax+dT, dT)

# Hamiltonian
H = np.matrix([[1, 0, 0, 0],
               [0, -1, 2, 0,],
               [0, 2, -1, 0],
               [0, 0, 0, 1]])
# Gate of one step
UdT = linalg.expm(-1j*H*dT)

# Target gate
Uswap = np.matrix([[1, 0, 0, 0], 
                   [0, 0, 1, 0],
                   [0, 1, 0, 0],
                   [0, 0, 0, 1]])

# Initiate variables
T = 0
U = np.identity(4)         
index = 0
InFidelity = np.zeros(int(np.floor(Tmax/dT))+1)

# Loop over durations T
for T in Tvector:
  Product = np.matmul(np.conj(Uswap.T), U)
  InFidelity[index] = 1 - np.abs(1/4*np.trace(Product))**2
  # Update gate and index
  U = np.matmul(UdT, U)
  index = index+1

# Plot cost function
plt.figure(1)
plt.clf()
plt.plot(Tvector, InFidelity, '-', linewidth = 2, color = 'black')
# Indicate exact zero-points
plt.vlines(np.pi/4, 0, 1, linewidth = 1.5, color = 'red', 
           linestyles = 'dashed')
plt.vlines(3*np.pi/4, 0, 1, linewidth = 1.5, color = 'red', 
           linestyles = 'dashed')
# Labels, fontsize and grid
plt.xlabel('Interaction duration T', fontsize = 12)
plt.ylabel('C(T)', fontsize = 12)
plt.grid()
plt.show()

File: MatlabCode/Chapter6_QuantumTech/Exercise_6_4_6.m
% This script sets out to fix the duration of the spin-spin interaction 
% between to spin 1/2-particles so that it, in effect, implements the 
% SWAP gate. The parameters are rigged such that the Hamiltonian assumes a 
% particularly simple form. The only inputs are the upper limit for the 
% interaction time T - and dT, the resolution in the vector with 
% time-durations.
% 
% It plots the cost function
% C(T) = 1 - |1/4 * Tr(U_target^\dagger U(T))|^2
% as a function of duration T; when C=0, our gate coincides
% with the target gate, and the fidelity is 100%.

% Inputs
Tmax = 3;
dT = 0.01;

% Vector with durations
Tvector = 0:dT:Tmax;

% Hamiltonian
H = [1 0 0 0; 0 -1 2 0; 0 2 -1 0; 0 0 0 1];
% Gate of one step
UdT = expm(-i*H*dT);

% Target gate
Uswap = [1 0 0 0; 0 0 1 0; 0 1 0 0; 0 0 0 1];

% Initiate variables
T = 0;
U = eye(4);         
index = 1;
InFidelity = zeros(1,floor(Tmax/dT));
% 
% Loop over durations T
while T < Tmax
  InFidelity(index) = 1 - abs(1/4*trace(Uswap'*U))^2;
  % Update gate, duration and index
  U = UdT*U;
  T = T+dT;
  index = index+1;
end

% Plot cost function
figure(1)
plot(Tvector, InFidelity, 'k-', 'linewidth', 2)
hold on
% Indicate exact zero-points
xline(pi/4, 'r--', 'linewidth', 1.5);
xline(3*pi/4, 'r--', 'linewidth', 1.5);
hold off
% Labels, fontsize and grid
xlabel('Interaction duration T')
ylabel('C(T)')
grid on
set(gca, 'fontsize', 15)
